### PR TITLE
[TE] Setup standalone mode for Summary algorithm

### PR DESCRIPTION
**Pull request by cyenjung**
1. Add a standalone mode for triggering the summary algorithm through command line.
2. Users could specify which cost function to be used through command line's arguments.
3. Refactor large portion of code and add unit tests.

Tested with new unit tests.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(className), "Class name of cost function cannot be empty.");
+
+    Class<CostFunction> clazz = (Class<CostFunction>) Class.forName(className);
+    Constructor<CostFunction> constructor = clazz.getConstructor(Map.class);
+    CostFunction costFunction = constructor.newInstance(new Object[] { params });
+
+    return costFunction;
+  }
+
+  public static void main(String[] args) throws Exception {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java]: Is this intended as a stand-alone CLI, or as a service?

Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java]: Yes, the stand-alone CLI is created for our users to conduct quick experiments without bringing up the entire ThirdEye.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/DimNameValueCostEntry.java [hunk 2]
```diff
@@ -1,20 +1,27 @@
...
   private double currentValue;
   private double baselineValue;
 
-  public DimNameValueCostEntry(String dimension, String dimValue, double dimValueCost, double contributionFactor, double currentValue, double baselineValue) {
-    this.dimName = dimension;
-    this.dimValue = dimValue;
-    this.cost = dimValueCost;
-    this.contributionFactor = contributionFactor;
-    this.currentValue = currentValue;
+  public DimNameValueCostEntry(String dimensionName, String dimensionValue, double baselineValue, double currentValue,
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/DimNameValueCostEntry.java]: why not pass the globalBaseline and globalCurrent, you can have methods such as getContributionFactor etc

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/Dimensions.java [hunk 3]
```diff
@@ -30,20 +33,66 @@ public String get(int index) {
...
-    return Collections.<String> emptyList();
+  /**
+   * Checks if the current dimension is the parent to the given dimension. A dimension A is a parent to dimension B if
+   * and only if dimension A is a subset of dimension B.
+   *
+   * @param child the given child dimension.
+   *
+   * @return true if the current dimension is a parent (subset) to the given dimension.
+   */
+  public boolean isParentTo(Dimensions child) {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/Dimensions.java]: isParentOf

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+      Summary summary = new Summary(cube, costFunction);
+      response = summary.computeSummary(summarySize, doOneSideError);
+    }
+    response.setDataset(dataset);
+    response.setMetricName(metric);
+
+    return response;
+  }
+
+  private static Options buildOptions() {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java]: move this to a separate class, call it ContributionAnalysisTool

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
+package com.linkedin.thirdeye.client.diffsummary.costfunction;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import java.util.Map;
+
+public class BalancedCostFunction implements CostFunction {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java]: add java doc explaining the high level idea

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+   * @param baselineValue the baseline value of the current node.
+   * @param currentValue the current value of the current node.
+   * @param parentRatio the ratio between baseline and current value of parent node.
+   * @param globalBaselineValue the baseline value of the top node.
+   * @param globalCurrentValue the current value of the top node.
+   *
+   * @return the cost that consider change difference, change ratio, and node size.
+   */
+  @Override
+  public double getCost(double baselineValue, double currentValue, double parentRatio, double globalBaselineValue,
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java]: computeCost is more appropriate

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/CostFunction.java [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.linkedin.thirdeye.client.diffsummary.costfunction;
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/CostFunction.java]: if this package contains different types of cost functions, the package should be plural. costfunctions or just functions

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/Dimensions.java (hunk 3 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/Dimensions.java]: Fixed.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java (hunk 4 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java]: Done!

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java (hunk 6 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java]: Done!

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/CostFunction.java (hunk 7 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/CostFunction.java]: Done!

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java (hunk 5 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/costfunction/BalancedCostFunction.java]: Done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/DimNameValueCostEntry.java (hunk 2 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/DimNameValueCostEntry.java]: There are two reasons that I didn't pass globalBaseline and globalCurrent to this class:
1. This class is used to store the information to be presented at the frontend, so it shouldn't contain any computational logic. The contribution factor should be calculated and pass to the instance of this class.
2. The value of  globalBaseline and globalCurrent are globally available in the summary that is pass to the front end. So the global values that are stored in this class will not be used in anyway.

**Pull request merged**